### ISSUE TITLE
Feature Request: Enhancement for 3.0

### ISSUE THREAD

**Issue by alanef** [Prospress/action-scheduler#341]
Hi,

I see that 3.0 will use its own tables.

can you implement a generic helper function like count_outstanding_by_group( $group )

this enables a program to not schedule until all jobs in a group are completed.

This is how I implemented it within the CPT post structure ( I'd even written a comment about storage change : )

Happy to pull 3.0 and write if you like

```
/**
	 * Counts action schedule jobs by group
	 * possible extension to ActionScheduler
	 * note would need to change if different storage used
	 *
	 * @param $group
	 *
	 * @return string|null
	 */
	public static function count_outstanding_by_group( $group ) {
		global $wpdb;
		$sql = "SELECT COUNT(DISTINCT p.ID) FROM {$wpdb->posts} p 
INNER JOIN $wpdb->term_relationships AS tr ON (p.ID = tr.object_id)
INNER JOIN $wpdb->term_taxonomy AS tt ON (tr.term_taxonomy_id = tt.term_taxonomy_id)
INNER JOIN $wpdb->terms AS t ON (t.term_id = tt.term_id)
WHERE   p.post_status IN ('in-progress','pending')
AND p.post_type = 'scheduled-action'
AND tt.taxonomy = %s 
AND t.name = %s";
		$sql = $wpdb->prepare( $sql, 'action-group', $group );

		return $wpdb->get_var( $sql );
	}
```

**Comment by rrennick** [comment]
This would fit pretty well in #279.

**Comment by alanef** [comment]
HAving had a poke around

It seems the Store classes have

`public function query_actions( $query = [], $query_type = 'select' ) {`

yet the abstract has 

`abstract public function query_actions( $query = array() );`

and the functions

```
function as_get_scheduled_actions( $args = array(), $return_format = OBJECT ) {
	$store = ActionScheduler::store();
	foreach ( array('date', 'modified') as $key ) {
		if ( isset($args[$key]) ) {
			$args[$key] = as_get_datetime_object($args[$key]);
		}
	}
	$ids = $store->query_actions( $args );
```

it would seem to me that simply adding the optional arg to the function ( and abstract ) for count would allow a function call to count by group - job done ( and could be used by CLI etc )

as in

```
function as_get_scheduled_actions( $args = array(), $return_format = OBJECT, $select_or_count = 'select' ) {
	$store = ActionScheduler::store();
	foreach ( array('date', 'modified') as $key ) {
		if ( isset($args[$key]) ) {
			$args[$key] = as_get_datetime_object($args[$key]);
		}
	}
	$ids = $store->query_actions( $args , $select_or_count);
```

I haven't tested my theory, but it seems viable - what are your thoughts?

**Comment by rrennick** [comment]
> what are your thoughts?

I generally prefer having a `*_get_{items}` function return [items] and a `*_get_{items}_count` function to return the count of the items. Sometimes, code brevity will be reason to consolidate. In this case it's only a few lines of code & I would do it as a separate function.

**Comment by alanef** [comment]
as in

```function as_get_scheduled_actions_count( $args = array(), $return_format = OBJECT)``

**Comment by rrennick** [comment]
`, $return_format = OBJECT` wouldn't be necessary as it's always returning an `int`.

**Comment by thenbrent** [comment]
> simply adding the optional arg to the function ( and abstract ) for count would allow a function call to count by group

I'm not sure why the abstract `ActionScheduler_Store::query_actions()` doesn't include the 2nd param. Possibly that was just to avoid breaking backward compatibility in the minor version it was released in. Both `ActionScheduler_wpPostStore` and `ActionScheduler_DBStore` support the 2nd param on `query_actions()` (as does the `ActionScheduler_HybridStore`).

The 2nd param is also used in core. For example, `ActionScheduler_ListTable::prepare_items()` calls `$this->store->query_actions( $query, 'count' )`, so really, it should be required.

Fortunately, as we're at a major version, we can include that param now, so I've submitted #342 to do that.

I don't think it's necessary to update add a new `as_get_scheduled_actions_count()` function (or add a new param to `as_get_scheduled_actions()`), because it's not a super common use case, and it would likely end up just containing this:

```php
return ActionScheduler::store()->query_actions( $args );
```

Happy to add that if there are reasons I'm missing though. It can also easily be adding in a future minor release as needed.

**Comment by alanef** [comment]
Thanks.

I'm surprised that it is not required more often.

My use case is primarily to email a summary report when all tasks in a group are complete and I would have thought multiple task batch dependency  would be more common. 

How close it 3.0.0 to being stable?

**Comment by alanef** [comment]
dont worry answering I have got that https://github.com/Prospress/action-scheduler/milestones

**Comment by alanef** [comment]
FYI

working for me in 3.0.0 data stores as follows

```/**
	 * Counts action schedule jobs by group
	 * for pending and in-progress so a job can run when all complete or failed
	 * 
	 * @param $group
	 *
	 * @return array
	 */
	public static function count_outstanding_by_group( $group ) {
		$args           = array(
			'group'    => $group,
			'per_page' => - 1,
			'status'   => 'pending',
		);
		$pending        = ActionScheduler::store()->query_actions( $args, 'count' );
		$args['status'] = 'in-progress';
		$in_progress    = ActionScheduler::store()->query_actions( $args, 'count' );

		return $pending + $in_progress;
	}
```

**Comment by thenbrent** [comment]
> dont worry answering I have got that https://github.com/Prospress/action-scheduler/milestones

3.0.0 is actually closer than that was showing yesterday. The only outstanding issue (other than this one) is #340.

**Comment by thenbrent** [comment]
Fixed with #342.
